### PR TITLE
Webpack, React and Preview2, oh my!

### PR DESCRIPTION
**Pull request by analogrelay**
* converted templating to use react
* use webpack for bundling and module management
* slight css restyling
* update to 2.0.0-preview2-final

There are a lot of deletions from `wwwroot/lib` because with WebPack, we pull them from `node_modules` as part of build now.

### CODE REVIEW

File: src/Hubbup.Web/Client/Standup/components.tsx [hunk 1]
```diff
@@ -0,0 +1,370 @@
+import * as React from 'react';
+
+import * as Data from '../data';
+
+const fetchSettings: RequestInit = {
+    credentials: "same-origin",
+    headers: {
+        ["X-Requested-With"]: "XMLHttpRequest"
```
Comment (SteveSandersonMS) [inline src/Hubbup.Web/Client/Standup/components.tsx]: Really minor point, but the use of `"` is inconsistent with previous lines 1 and 3. Most React devs tend to use single quotes everywhere except (for some reason I've never understood) around HTML attribute values where they tend to use double-quotes.

Comment (SteveSandersonMS) [reply src/Hubbup.Web/Client/Standup/components.tsx]: Also you don't need the square brackets around `"X-Requested-With"` since they are only needed when evaluating dynamic expressions.

File: src/Hubbup.Web/Client/Standup/components.tsx [hunk 2]
```diff
@@ -0,0 +1,370 @@
...
+    render() {
+        return <a href={this.props.user.url}>
+            <img style={{ "margin": "0.1em" }} height="32" width="32" src={this.props.user.avatarUrl} title={this.props.user.login} alt={this.props.user.login} />
+        </a>;
+    }
+}
+
+class IssueUserView extends React.Component<{ issue: Data.Issue }, undefined> {
+    render() {
+        let issue = this.props.issue;
```
Comment (SteveSandersonMS) [inline src/Hubbup.Web/Client/Standup/components.tsx]: Another minor point, but it's more conventional to use `const` rather than `let` by default, and only use `let` for the special cases where you need to emphasise that the variable will get written to after definition.

File: src/Hubbup.Web/Client/Standup/components.tsx [hunk 3]
```diff
@@ -0,0 +1,370 @@
...
+        let issueLink = <span className="issue-link">
+            <a href={issue.url}>
+                #{issue.number}
+            </a>
+        </span>;
+
+        let style = { "whiteSpace": "nowrap" };
+        if (issue.assignees && issue.assignees[0] && issue.assignees[0].id != issue.author.id) {
```
Comment (SteveSandersonMS) [inline src/Hubbup.Web/Client/Standup/components.tsx]: Another stylistic point, but this one's a bit more important - always use strict equality/inequality operators (in this case, `!==` rather than `!=`) unless for some incredibly rare reason you actually want the crazy semantics of loose equality/inequality, and in that case it warrants a comment to explain why loose equality/inequality is the relevant behaviour.

File: src/Hubbup.Web/Client/Standup/components.tsx [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+                <span className="badge">{this.props.issues.length}</span>
+                {' ' /* https://facebook.github.io/react/blog/2014/02/20/react-v0.9.html#jsx-whitespace */}
+                {collapser}
+            </h4>
+            {content}
+        </div>;
+    }
+}
+
+type RateLimitUpdater = (graphQl: { cost: number, remaining: number, resetAt: string }, rest: { remaining: number, reset: string }) => void;
```
Comment (SteveSandersonMS) [inline src/Hubbup.Web/Client/Standup/components.tsx]: Seems a bit random that this is an inline `type` rather than an `interface` like the others below, but I guess it will produce the same effect.

Comment (SteveSandersonMS) [reply src/Hubbup.Web/Client/Standup/components.tsx]: Oh I see it's a function. You can do it with an `interface` like this: https://www.typescriptlang.org/docs/handbook/interfaces.html#function-types - but whether that's clearer or not is subjective.

File: src/Hubbup.Web/Client/Standup/components.tsx [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+        });
+
+        let resp = await fetch(`${this.props.baseUrl}api/repoSets/${this.props.repoSet}/issues/${this.props.login}`, fetchSettings)
+        if (resp.status < 200 || resp.status > 299) {
+            this.setState({
+                loading: false,
+                error: `Unexpected response fetching issues: ${resp.status} ${resp.statusText}`,
+                data: null
+            });
+            return;
```
Comment (SteveSandersonMS) [inline src/Hubbup.Web/Client/Standup/components.tsx]: Unnecessary line

File: src/Hubbup.Web/Client/Standup/components.tsx [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+    async reload() {
+        // This doesn't immediately clear the loading widget, but it will when the function returns and we re-render.
+        this.setState({
+            loading: true,
+            error: "",
+            data: null
+        });
+
+        let resp = await fetch(`${this.props.baseUrl}api/repoSets/${this.props.repoSet}/issues/${this.props.login}`, fetchSettings)
+        if (resp.status < 200 || resp.status > 299) {
```
Comment (SteveSandersonMS) [inline src/Hubbup.Web/Client/Standup/components.tsx]: Could replace this check with `if (!resp.ok) {`. Here's the spec for that property: https://github.com/whatwg/fetch/commit/e62f57b6c923f16e877ab56913a2a65388a7535a

File: src/Hubbup.Web/Client/Standup/components.tsx [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+
+    async reload() {
+        // This doesn't immediately clear the loading widget, but it will when the function returns and we re-render.
+        this.setState({
+            loading: true,
+            error: "",
+            data: null
+        });
+
+        let resp = await fetch(`${this.props.baseUrl}api/repoSets/${this.props.repoSet}/issues/${this.props.login}`, fetchSettings)
```
Comment (SteveSandersonMS) [inline src/Hubbup.Web/Client/Standup/components.tsx]: Should also have a `try`/`catch` around this, because the `fetch` promise can also be rejected if there's a network error (as opposed to if you get an HTTP response with some status like 500, where the promise succeeds but `resp.ok` is false).

File: src/Hubbup.Web/Client/Standup/components.tsx [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+            "marginLeft": "0.5em",
+            "marginTop": "-5px"
+        };
+
+        if (this.state.loading) {
+            content = <div className="panel-body">
+                <LoadingPanel />
+            </div>;
+        }
+        else if (this.state.error) {
```
Comment (SteveSandersonMS) [inline src/Hubbup.Web/Client/Standup/components.tsx]: More nitpicking :) The brace style is weird here. Normally you'd see `} else if (...) {` all on one line.

File: src/Hubbup.Web/Client/Standup/components.tsx [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+        let newState = { ... this.state };
+        let resp = await fetch(`${this.props.baseUrl}api/repoSets/${this.props.repoSet}/people`, fetchSettings);
+        if (resp.status < 200 || resp.status > 299) {
+            newState.loading = false;
+            newState.error = `Unexpected response from server: ${resp.status} ${resp.statusText}`;
+        }
+        else {
+            newState.loading = false;
+            newState.error = "";
+            newState.people = (await resp.json()) as string[];
```
Comment (SteveSandersonMS) [inline src/Hubbup.Web/Client/Standup/components.tsx]: TBH it's a bit unusual to see all these object-mutating assignments in a React app. Most React devs find that functional programming habits mesh better with React's philosophy. I know it's a minor difference, but it would be more typical to phrase this logic more like:

```
    this.setState({
        ...this.state, // Instead of line 305
        loading: false,
        error: '',
        people: await resp.json() // Not sure you strictly need the `as` cast, if the state type already has a type for `people`
    });
```

This is more just a thought to consider rather than something you should necessarily change, since of course it doesn't actually affect behaviour.

File: src/Hubbup.Web/Client/Standup/components.tsx [hunk 10]
```diff
@@ -0,0 +1,370 @@
...
+                rest: null,
+                sinceReload: 0
+            }
+        };
+    }
+
+    async componentDidMount() {
+        // Fetch the list of people
+        let newState = { ... this.state };
+        let resp = await fetch(`${this.props.baseUrl}api/repoSets/${this.props.repoSet}/people`, fetchSettings);
```
Comment (SteveSandersonMS) [inline src/Hubbup.Web/Client/Standup/components.tsx]: Also needs `try`/`catch` and possibly use `resp.ok` below.

File: src/Hubbup.Web/Client/Standup/page.tsx [hunk 11]
```diff
@@ -0,0 +1,18 @@
+import * as React from 'react';
+import * as ReactDOM from 'react-dom';
+
+import * as Data from '../data';
+import { Page } from './components';
+
+document.addEventListener("DOMContentLoaded", async () => {
```
Comment (SteveSandersonMS) [inline src/Hubbup.Web/Client/Standup/page.tsx]: Don't think `async` is needed here

File: src/Hubbup.Web/Pages/Index.cshtml [hunk 12]
```diff
@@ -1,312 +1,19 @@
...
-            <div class="tab-pane active" id="group-by-assignee">
-                @foreach (var person in Model.People)
-                {
-                    <div class="panel panel-default i-assignee-panel" data-assignee="@person" data-id="@TagBuilder.CreateSanitizedId(person, "-")">
-                    </div>
-                }
-            </div>
+    <div class="col-md-12">
+        <a asp-controller="IssueList" asp-action="Index" asp-route-repoSet="@repoSet">Switch to Legacy View</a>
+        <div class="tab-content" id="app-root" data-repo-set="@repoSet" data-base-url="@Url.Content("~/")" data-environment="@HostingEnvironment.EnvironmentName">
```
Comment (SteveSandersonMS) [inline src/Hubbup.Web/Pages/Index.cshtml:12]: This is a nice pattern - using the root elem attributes to pass over data from the .NET side. I've more often seen people use global vars, but your approach here is cleaner.

File: src/Hubbup.Web/webpack.config.js [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+            name: 'vendor',
+            minChunks: function (module) {
+                // this assumes your vendor imports exist in the node_modules directory
+                return module.context && module.context.indexOf('node_modules') !== -1;
+            }
+        }),
+
+        function () {
+            this.plugin("done", function (stats) {
+                path.join()
```
Comment (SteveSandersonMS) [inline src/Hubbup.Web/webpack.config.js]: I guess this isn't meant to be here...

**Comment by SteveSandersonMS** [comment]
This looks great! I've posted a bunch of nitpicky style comments but the overall structure looks clean and easy to understand.

Couple of thoughts:

 * It looks like you don't have Webpack dev middleware / HMR set up, so I'm guessing you needed to use some extra command-line tooling to build and then manually reload after each change. We have Webpack dev middleware / HMR support in the box with SpaServices. If you want I could submit a PR to set this up, though if you want a more accurate assessment of the customer experience you might want to try doing that yourself.
 * I'm interested in your reflections on using React. It definitely appears that you made good use of `tsx` syntax to instantiate and pass around components, which hopefully was a better experience than you had before with handlebars. However the app is pretty much data-display-only (no editing) and with minimal if any navigation, so perhaps some of the wider benefits didn't apply in this case. Given the choice, would you build it this way again?

**Review by SteveSandersonMS — APPROVED** [review]
The only thing I think you really *should* change is using `try`/`catch` around any `await fetch` and switching to strict equality/inequality comparisons. All the other comments are just nice-to-haves. Nice job!

**Comment by SteveSandersonMS** [comment]
Also now I see how this works, if you want, we could add server-side prerendering. It would be a different pattern to what we use in the JavaScriptServices templates (because you're not using Redux), and the practical benefit would not be major (because this is not really targeting people using slow mobile networks or very slow devices), so it would be more of a tech demo than really mattering in this case.

**Comment by analogrelay** [comment]
> We have Webpack dev middleware / HMR support in the box with SpaServices.

Yeah, I looked at that a bit recently, but I didn't want to overload this PR with it for now :). That could be an interesting next step to take this in.

> I'm interested in your reflections on using React.

I've used React a little in the past, and I do feel like it's a little too complicated for this kind of app. I think it was a good experience to adapt it, but React seems more beneficial in apps that have significant state changing (editing, frequently updating data, etc.), and a little overkill for something that just renders some HTML from a JSON blob fetched once. I added per-user refresh just to kinda get a little bit of the benefit (even though that's not really a feature that's very important).

> I've posted a bunch of nitpicky style comments

I'll get all those resolved! I appreciate the style input, as I'm not really familiar with React idioms ;)

File: src/Hubbup.Web/Client/Standup/components.tsx (hunk 1 above)
Comment (analogrelay) [reply src/Hubbup.Web/Client/Standup/components.tsx]: I"m usually pretty 'inconsistent' with my quotes in JavaScript. It's kinda a pain that it"ll allow both kinds of quotes ;P.

Trolling aside, I'll do a quick pass to make them consistent. If I cared more, I'd set up a linter, but I think that can wait for now.

File: src/Hubbup.Web/Client/Standup/components.tsx (hunk 3 above)
Comment (analogrelay) [reply src/Hubbup.Web/Client/Standup/components.tsx]: Yep, I generally tried to use strict comparison, but I forgot here and didn't have a linter set up :).

File: src/Hubbup.Web/Client/Standup/components.tsx (hunk 4 above)
Comment (analogrelay) [reply src/Hubbup.Web/Client/Standup/components.tsx]: Actually this needn't be a `type` or `interface` at all. I added it when I thought I was going to use it in multiple places but I didn't end up doing so.

File: src/Hubbup.Web/Client/Standup/page.tsx (hunk 11 above)
Comment (analogrelay) [reply src/Hubbup.Web/Client/Standup/page.tsx]: Yep, it's vestigial.

**Comment by analogrelay** [comment]
:up: :date: should have resolved all the feedback now!

**Comment by SteveSandersonMS** [comment]
Looks extra great now!

File: src/Hubbup.Web/launchSettings.json [hunk 14]
```diff
@@ -0,0 +1,27 @@
+{
```
Comment (Eilon) [inline src/Hubbup.Web/launchSettings.json]: Is this file needed in addition to https://github.com/Eilon/ProjectKIssueList/blob/master/src/Hubbup.Web/Properties/launchSettings.json ?

**Review by Eilon — CHANGES_REQUESTED** [review]

File: src/Hubbup.Web/launchSettings.json (hunk 14 above)
Comment (analogrelay) [reply src/Hubbup.Web/launchSettings.json]: No... and I'm a little unsure where it came from. Removing it now.

Comment (analogrelay) [reply src/Hubbup.Web/launchSettings.json]: Resolved.
